snapcraft/internal/meta/_desktop.py

```diff
@@ -56,7 +57,15 @@ def parse_and_reformat(self):
             exec_value = "{}.{} %U".format(self._snap_name, self._name)
         self._parser[section]["Exec"] = exec_value
         if "Icon" in self._parser[section]:
-            icon = self._parser[section]["Icon"]
+            # Extracted metadata (e.g. from the AppStream) can override the
+            # icon location.
+            if extracted_metadata:
+                metadata_dict = extracted_metadata.to_dict()
```

Comment:
why not use `get_icon()`? It should return None if not found or set.

Reply:
Changed to use get_icon(), but also refactored the override logic to prevent icon to remain unassigned.